### PR TITLE
Add CMake build for C++ and scikit-build infrastructure for Python package installation

### PR DESCRIPTION
**Pull request by jjwilke**
Clone of https://github.com/nv-legate/cunumeric/pull/311 with special branch name so CI will test the legate.core and cunumeric PR's together.

### CODE REVIEW

**Review by trxcllnt — APPROVED** [review]

File: cunumeric_cpp.cmake [hunk 1]
```diff
@@ -355,6 +355,7 @@ target_link_libraries(cunumeric
           $<TARGET_NAME_IF_EXISTS:conda_env>
           $<TARGET_NAME_IF_EXISTS:CUDA::cufft>
           $<TARGET_NAME_IF_EXISTS:CUDA::cublas>
+          $<TARGET_NAME_IF_EXISTS:CUDA::curand>
```
Comment (magnatelee) [inline cunumeric_cpp.cmake:358]: it looks like we don't need to link against libcurand (see https://github.com/nv-legate/cunumeric/pull/541). can you try to build things without this line and see if that works?

File: conda/environment-test-3.9.yml [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
   # build
-  - cutensor>=1.3.3
+  - git
   - nccl
+  - make
+  - zlib
+  - cmake
+  - ninja
+  - c-compiler
+  - cxx-compiler
```
Comment (bryevdv) [inline conda/environment-test-3.9.yml:15]:
```suggestion
  - cmake>=3.24
  - ninja
  - c-compiler
  - cxx-compiler
  - gcc_linux-64=9.* # [linux64]
  - sysroot_linux-64==2.17 # [linux64]
```

File: conda/environment-test-3.8.yml [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
   # build
-  - cutensor>=1.3.3
+  - git
   - nccl
+  - make
+  - zlib
+  - cmake
+  - ninja
+  - c-compiler
+  - cxx-compiler
```
Comment (bryevdv) [inline conda/environment-test-3.8.yml:15]:
```suggestion
  - cmake>=3.24
  - ninja
  - c-compiler
  - cxx-compiler
  - gcc_linux-64=9.* # [linux64]
  - sysroot_linux-64==2.17 # [linux64]
```

File: conda/environment-test-3.10.yml [hunk 4]
```diff
@@ -1,14 +1,21 @@
...
   # build
-  - cutensor>=1.3.3
+  - git
   - nccl
+  - make
+  - zlib
+  - cmake
+  - ninja
+  - c-compiler
+  - cxx-compiler
```
Comment (bryevdv) [inline conda/environment-test-3.10.yml:15]:
```suggestion
  - cmake>=3.24
  - ninja
  - c-compiler
  - cxx-compiler
  - gcc_linux-64=9.* # [linux64]
  - sysroot_linux-64==2.17 # [linux64]
```

File: conda/environment-test-3.8.yml [hunk 5]
```diff
@@ -1,14 +1,23 @@
...
-  - cutensor>=1.3.3
+  - git
   - nccl
+  - make
+  - zlib
+  - cmake>=3.24
+  - ninja
+  - c-compiler
+  - cxx-compiler
+  - gcc_linux-64=9.* # [linux64]
```
Comment (bryevdv) [inline conda/environment-test-3.8.yml:16]:
```suggestion
  - gcc_linux-64 # [linux64]
```

File: conda/environment-test-3.9.yml [hunk 6]
```diff
@@ -1,14 +1,23 @@
...
-  - cutensor>=1.3.3
+  - git
   - nccl
+  - make
+  - zlib
+  - cmake>=3.24
+  - ninja
+  - c-compiler
+  - cxx-compiler
+  - gcc_linux-64=9.* # [linux64]
```
Comment (bryevdv) [inline conda/environment-test-3.9.yml:16]:
```suggestion
  - gcc_linux-64 # [linux64]
```

File: cunumeric_cpp.cmake (hunk 1 above)
Comment (trxcllnt) [reply cunumeric_cpp.cmake:358]: Sure, I can remove it. Conda prints a warning that the `libcurand` package isn't used when I build the conda package locally, but I'm fine ignoring that message.

**Comment by marcinz** [comment]
@trxcllnt Are the recent changes in CI responsible for the test failure, or is it something else?

**Comment by trxcllnt** [comment]
@marcinz do we need to rebuilt the CI images after merging that PR earlier?

**Comment by marcinz** [comment]
So the problem is that we need the modified Github workflow from the conda build PRs.

File: install.py [hunk 7]
```diff
@@ -345,10 +345,7 @@ def validate_path(path):
     if cuda and curand_dir is not None:
         cmake_flags += ["-Dcunumeric_cuRAND_INCLUDE_DIR=%s" % curand_dir]
     if legate_dir:
-        # TODO: Undo this! Only setting temporarily so CI passes
-        if exists(join(legate_dir, "lib", "cmake", "legate_core")):
-            cmake_flags += ["-Dlegate_core_ROOT=%s" % legate_dir]
-        elif exists(join(legate_dir, "legate_core-config.cmake")):
+        if exists(join(legate_dir, "legate_core-config.cmake")):
```
Comment (manopapad) [inline install.py:348]: I think just
```
if legate_dir:
  cmake_flags += ["-Dlegate_core_ROOT=%s" % legate_dir]
```
We'll find out if we did it right once we see how CI works post-merge :-P

**Pull request merged**